uqcsbot/whatweekisit.py

```diff
@@ -0,0 +1,160 @@
+import requests
+from collections import namedtuple
+from typing import NamedTuple, List, Tuple, Optional
+from datetime import datetime, timedelta
+from math import ceil
+
+from bs4 import BeautifulSoup
+from discord.ext import commands
+
+from uqcsbot.utils.command_utils import loading_status
+
+# Endpoint that contains a table of semester dates
+MARKUP_CALENDAR_URL: str = "https://systems-training.its.uq.edu.au/systems/student-systems/electronic-course-profile-system/design-or-edit-course-profile/academic-calendar-teaching-week"
+DATE_FORMAT = "%d/%m/%Y"
+
+
+# Semester information
+class Semester(NamedTuple):
+    name: str
+    start_date: datetime
+    end_date: datetime
+    weeks: List[str]
+
+
+def date_to_string(date: datetime):
+    """
+    Formats date based on the format used for the command
+        Returns:
+            Stringified date
+    """
+    return date.strftime(DATE_FORMAT)
+
+
+def string_to_date(date: str) -> datetime:
+    """
+    Returns datetime object from string based on the format used for the command
+        Returns:
+            Stringified date
+    """
+    return datetime.strptime(date, DATE_FORMAT)
+
+
+def get_semester_times(markup: str) -> List[Semester]:
+    """
+    Parses the given HTML page to get a list of the names, dates & weeks of the semesters at UQ
+
+        Dev: this relies on the calendar page for UQ staying relative sane and static
+        Parameters:
+            markup: HTML page from UQ's website in plaintext
+        Returns:
+            A list of the information about semesters at UQ
+
+    """
+    soup = BeautifulSoup(markup, "html.parser")
+    semesters: List[Semester] = []
+
+    for semester_title in soup.find_all("h2", "structured-page__step-title"):
+        table = semester_title.parent.find("table")
+        weeks = table.find("tbody").find_all("tr")
+
+        # We assume the term starts on a Monday
+        sem_start = weeks[0].find_all("td")[0].text
+        start_date: datetime = string_to_date(sem_start)
+
+        # We'll filter out the empty cells at the end of the term, say in case it ends on a Tuesday
+        last_week_dates = [
+            cell for cell in weeks[-1].find_all("td") if len(cell.text.strip())
+        ]
+        end_date: datetime = string_to_date(last_week_dates[-1].text)
+
+        week_names = [week.text.strip() for week in table.find("tbody").find_all("th")]
+        # Invariant
+        assert len(week_names) == ceil(
+            (end_date - start_date + timedelta(days=1)).days / 7
+        )
+
+        semesters.append(
+            Semester(semester_title.text, start_date, end_date, week_names)
+        )
+
+    return semesters
+
+
+def get_semester_week(
+    semesters: List[Semester], checked_date: datetime
+) -> Tuple[Optional[str], Optional[str], Optional[str]]:
```

Comment:
types: the later code checks if semester (the first field) is not None then assumes everything else is not None.

hence, a more precise type would be:
> Optional[Tuple[str,str,str]]

Reply:
Yeah oop, I was being lazy with that -- I changed it so that it returns `None` if it can't find a match and the valid tuple otherwise